sand/controllers/transform.py

```diff
@@ -0,0 +1,190 @@
+            tdata["error"] = "".join(traceback.format_exception(exc, value, tb))
+            transformed_data.append(tdata)
+            exec_tolerance -= 1
+            if exec_tolerance == 0:
+                return transformed_data
+    return transformed_data
+
+
+def transform_split(transform_func, data, tolerance):
+    transformed_data = []
+    exec_tolerance = tolerance
+    for path, value, context in data:
+        tdata = dict()
+        value = value[0]
+        tdata["path"] = path
+        tdata["value"] = value
+        try:
+            result = transform_func(value, context)
+            if type(result) != list:
+                raise BadRequest("filter transform function must return list")
+            tdata["ok"] = result
+            transformed_data.append(tdata)
+        except Exception as err:
+            (exc, value, tb) = sys.exc_info()
+            tb = tb.tb_next  # Skip *this* frame
+            tdata["error"] = "".join(traceback.format_exception(exc, value, tb))
+            transformed_data.append(tdata)
+            exec_tolerance -= 1
+            if exec_tolerance == 0:
+                return transformed_data
+    return transformed_data
+
+
+def transform_concatenate(transform_func, data, tolerance):
+    transformed_data = []
+    exec_tolerance = tolerance
+    for path, value, context in data:
+        tdata = dict()
+        tdata["path"] = path
+        tdata["value"] = value
+        try:
+            result = transform_func(value, context)
+            tdata["ok"] = result
+            transformed_data.append(tdata)
+        except Exception as err:
+            (exc, value, tb) = sys.exc_info()
+            tb = tb.tb_next  # Skip *this* frame
+            tdata["error"] = "".join(traceback.format_exception(exc, value, tb))
+            transformed_data.append(tdata)
+            exec_tolerance -= 1
+            if exec_tolerance == 0:
+                return transformed_data
+    return transformed_data
+
+
+@transform_bp.route(
+    f"/{transform_bp.name}/<table_id>/transformations", methods=["POST"]
+)
+def transform(table_id: int):
+    table = Table.get_by_id(table_id)
+    table_rows: List[TableRow] = list(
+        TableRow.select().where(TableRow.table == table).order_by(TableRow.index)
+    )
+    transform_type = request.json["type"]
+    mode = request.json["mode"]
+    datapath = request.json["datapath"]
+    if type(datapath) != list and type(datapath) == str:
+        datapath = [datapath]
+
+    code = request.json["code"]
+    tolerance = request.json["tolerance"]
+    outputpath = None
+    if "outputpath" in request.json:
+        outputpath = request.json["outputpath"]
+    rows = request.json["rows"]
+
+    loc = {}
+
+    # overriding inbuilt _getitem_ guard function
+    def custom_getitem_guard(iterable, index):
+        return iterable[index]
+
+    # update safe_globals to enable _getitem_
+    safe_globals.update({'_getitem_': custom_getitem_guard})
```

Comment:
the function compilation gets complicated. move to a separate function

Reply:
implemented compile_function()